In [1]:
#Baseball player salary prediction model

In [2]:
# import the library

import pandas as pd

import numpy as np

import matplotlib.pyplot as plt

import seaborn as sns

print('imported')






imported


In [3]:
data= pd.read_csv('/content/hitters.csv')

In [4]:
data

Unnamed: 0  AtBat  Hits  HmRun  Runs  RBI  Walks  Years  CAtBat  \
0       -Andy Allanson    293    66      1    30   29     14      1     293   
1          -Alan Ashby    315    81      7    24   38     39     14    3449   
2         -Alvin Davis    479   130     18    66   72     76      3    1624   
3        -Andre Dawson    496   141     20    65   78     37     11    5628   
4    -Andres Galarraga    321    87     10    39   42     30      2     396   
..                 ...    ...   ...    ...   ...  ...    ...    ...     ...   
317      -Willie McGee    497   127      7    65   48     37      5    2703   
318   -Willie Randolph    492   136      5    76   50     94     12    5511   
319    -Wayne Tolleson    475   126      3    61   43     52      6    1700   
320     -Willie Upshaw    573   144      9    85   60     78      8    3198   
321     -Willie Wilson    631   170      9    77   44     31     11    4908   

     CHits  ...  CRuns  CRBI  CWalks  League Division PutOuts  Assists  \
0       66  ...     30    29      14       A        E     446       33   
1      835  ...    321   414     375       N        W     632       43   
2      457  ...    224   266     263       A        W     880       82   
3     1575  ...    828   838     354       N        E     200       11   
4      101  ...     48    46      33       N        E     805       40   
..     ...  ...    ...   ...     ...     ...      ...     ...      ...   
317    806  ...    379   311     138       N        E     325        9   
318   1511  ...    897   451     875       A        E     313      381   
319    433  ...    217    93     146       A        W      37      113   
320    857  ...    470   420     332       A        E    1314      131   
321   1457  ...    775   357     249       A        W     408        4   

     Errors  Salary  NewLeague  
0        20     NaN          A  
1        10   475.0          N  
2        14   480.0          A  
3         3   500.0          N  
4         4    91.5          N  
..      ...     ...        ...  
317       3   700.0          N  
318      20   875.0          A  
319       7   385.0          A  
320      12   960.0          A  
321       3  1000.0          A  

[322 rows x 21 columns]

A data frame with 322 observations of major league players on the following 20 variables.

AtBat Number of times at bat in 1986

Hits Number of hits in 1986

HmRun Number of home runs in 1986

Runs Number of runs in 1986

RBI Number of runs batted in in 1986

Walks Number of walks in 1986

Years Number of years in the major leagues

CAtBat Number of times at bat during his career

CHits Number of hits during his career

CHmRun Number of home runs during his career

CRuns Number of runs during his career

CRBI Number of runs batted in during his career

CWalks Number of walks during his career

League A factor with levels A and N indicating player’s league at the end of 1986

Division A factor with levels E and W indicating player’s division at the end of 1986

PutOuts Number of put outs in 1986

Assists Number of assists in 1986

Errors Number of errors in 1986

Salary 1987 annual salary on opening day in thousands of dollars NewLeague A factor with levels A and N indicating player’s league at the beginning of 1987









In [5]:
# check the columns
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 322 entries, 0 to 321
Data columns (total 21 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Unnamed: 0  322 non-null    object 
 1   AtBat       322 non-null    int64  
 2   Hits        322 non-null    int64  
 3   HmRun       322 non-null    int64  
 4   Runs        322 non-null    int64  
 5   RBI         322 non-null    int64  
 6   Walks       322 non-null    int64  
 7   Years       322 non-null    int64  
 8   CAtBat      322 non-null    int64  
 9   CHits       322 non-null    int64  
 10  CHmRun      322 non-null    int64  
 11  CRuns       322 non-null    int64  
 12  CRBI        322 non-null    int64  
 13  CWalks      322 non-null    int64  
 14  League      322 non-null    object 
 15  Division    322 non-null    object 
 16  PutOuts     322 non-null    int64  
 17  Assists     322 non-null    int64  
 18  Errors      322 non-null    int64  
 19  Salary      263 non-null    f

In [6]:
# look for missing value


In [7]:
data.isnull().sum()

Unnamed: 0     0
AtBat          0
Hits           0
HmRun          0
Runs           0
RBI            0
Walks          0
Years          0
CAtBat         0
CHits          0
CHmRun         0
CRuns          0
CRBI           0
CWalks         0
League         0
Division       0
PutOuts        0
Assists        0
Errors         0
Salary        59
NewLeague      0
dtype: int64

In [8]:
#drop the col unnamed as it is not required

data.drop('Unnamed: 0',axis=1,inplace=True)






In [9]:
data.head()






AtBat  Hits  HmRun  Runs  RBI  Walks  Years  CAtBat  CHits  CHmRun  CRuns  \
0    293    66      1    30   29     14      1     293     66       1     30   
1    315    81      7    24   38     39     14    3449    835      69    321   
2    479   130     18    66   72     76      3    1624    457      63    224   
3    496   141     20    65   78     37     11    5628   1575     225    828   
4    321    87     10    39   42     30      2     396    101      12     48   

   CRBI  CWalks League Division  PutOuts  Assists  Errors  Salary NewLeague  
0    29      14      A        E      446       33      20     NaN         A  
1   414     375      N        W      632       43      10   475.0         N  
2   266     263      A        W      880       82      14   480.0         A  
3   838     354      N        E      200       11       3   500.0         N  
4    46      33      N        E      805       40       4    91.5         N

In [10]:
# fill missing values

data['Salary'].fillna(data['Salary'].median(),inplace=True)






/tmp/ipython-input-323792637.py:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  data['Salary'].fillna(data['Salary'].median(),inplace=True)


In [11]:
# check again
data.isnull().sum()

AtBat        0
Hits         0
HmRun        0
Runs         0
RBI          0
Walks        0
Years        0
CAtBat       0
CHits        0
CHmRun       0
CRuns        0
CRBI         0
CWalks       0
League       0
Division     0
PutOuts      0
Assists      0
Errors       0
Salary       0
NewLeague    0
dtype: int64

In [12]:
# describe()
data.describe()


AtBat        Hits       HmRun        Runs         RBI       Walks  \
count  322.000000  322.000000  322.000000  322.000000  322.000000  322.000000   
mean   380.928571  101.024845   10.770186   50.909938   48.027950   38.742236   
std    153.404981   46.454741    8.709037   26.024095   26.166895   21.639327   
min     16.000000    1.000000    0.000000    0.000000    0.000000    0.000000   
25%    255.250000   64.000000    4.000000   30.250000   28.000000   22.000000   
50%    379.500000   96.000000    8.000000   48.000000   44.000000   35.000000   
75%    512.000000  137.000000   16.000000   69.000000   64.750000   53.000000   
max    687.000000  238.000000   40.000000  130.000000  121.000000  105.000000   

            Years       CAtBat        CHits      CHmRun        CRuns  \
count  322.000000    322.00000   322.000000  322.000000   322.000000   
mean     7.444099   2648.68323   717.571429   69.490683   358.795031   
std      4.926087   2324.20587   654.472627   86.266061   334.105886   
min      1.000000     19.00000     4.000000    0.000000     1.000000   
25%      4.000000    816.75000   209.000000   14.000000   100.250000   
50%      6.000000   1928.00000   508.000000   37.500000   247.000000   
75%     11.000000   3924.25000  1059.250000   90.000000   526.250000   
max     24.000000  14053.00000  4256.000000  548.000000  2165.000000   

              CRBI       CWalks      PutOuts     Assists      Errors  \
count   322.000000   322.000000   322.000000  322.000000  322.000000   
mean    330.118012   260.239130   288.937888  106.913043    8.040373   
std     333.219617   267.058085   280.704614  136.854876    6.368359   
min       0.000000     0.000000     0.000000    0.000000    0.000000   
25%      88.750000    67.250000   109.250000    7.000000    3.000000   
50%     220.500000   170.500000   212.000000   39.500000    6.000000   
75%     426.250000   339.250000   325.000000  166.000000   11.000000   
max    1659.000000  1566.000000  1378.000000  492.000000   32.000000   

            Salary  
count   322.000000  
mean    515.600953  
std     409.817459  
min      67.500000  
25%     226.250000  
50%     425.000000  
75%     700.000000  
max    2460.000000

In [13]:
# Apply machine learning model to predict salary


In [14]:
# data pre processing step
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 322 entries, 0 to 321
Data columns (total 20 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   AtBat      322 non-null    int64  
 1   Hits       322 non-null    int64  
 2   HmRun      322 non-null    int64  
 3   Runs       322 non-null    int64  
 4   RBI        322 non-null    int64  
 5   Walks      322 non-null    int64  
 6   Years      322 non-null    int64  
 7   CAtBat     322 non-null    int64  
 8   CHits      322 non-null    int64  
 9   CHmRun     322 non-null    int64  
 10  CRuns      322 non-null    int64  
 11  CRBI       322 non-null    int64  
 12  CWalks     322 non-null    int64  
 13  League     322 non-null    object 
 14  Division   322 non-null    object 
 15  PutOuts    322 non-null    int64  
 16  Assists    322 non-null    int64  
 17  Errors     322 non-null    int64  
 18  Salary     322 non-null    float64
 19  NewLeague  322 non-null    object 
dtypes: float64

In [15]:
# all the object data need to be converted into dummy type
data['League'].value_counts()

League
A    175
N    147
Name: count, dtype: int64

In [16]:
data['Division'].value_counts()

Division
W    165
E    157
Name: count, dtype: int64

In [17]:
# create dummy objects for object type data
# create dummy vars for object data

#dummy_data=pd.get_dummies(data[['League','Division','NewLeague']]).astype(int)

#dummy_data






In [18]:
# doing again
# create dummy vars for object data with drop first, because if two values then variable can decide, same for more

dummy_data=pd.get_dummies(data[['League','Division','NewLeague']],drop_first=True).astype(int)

dummy_data

# after drop first only one variable for each of the value






League_N  Division_W  NewLeague_N
0           0           0            0
1           1           1            1
2           0           1            0
3           1           0            1
4           1           0            1
..        ...         ...          ...
317         1           0            1
318         0           0            0
319         0           1            0
320         0           0            0
321         0           1            0

[322 rows x 3 columns]

In [19]:
#create dep var & indep var
# salary is as outcome variable and other categorical are also dropped

y=data['Salary']

X=data.drop(['League','Division','NewLeague','Salary'],axis=1)
data






AtBat  Hits  HmRun  Runs  RBI  Walks  Years  CAtBat  CHits  CHmRun  \
0      293    66      1    30   29     14      1     293     66       1   
1      315    81      7    24   38     39     14    3449    835      69   
2      479   130     18    66   72     76      3    1624    457      63   
3      496   141     20    65   78     37     11    5628   1575     225   
4      321    87     10    39   42     30      2     396    101      12   
..     ...   ...    ...   ...  ...    ...    ...     ...    ...     ...   
317    497   127      7    65   48     37      5    2703    806      32   
318    492   136      5    76   50     94     12    5511   1511      39   
319    475   126      3    61   43     52      6    1700    433       7   
320    573   144      9    85   60     78      8    3198    857      97   
321    631   170      9    77   44     31     11    4908   1457      30   

     CRuns  CRBI  CWalks League Division  PutOuts  Assists  Errors  Salary  \
0       30    29      14      A        E      446       33      20   425.0   
1      321   414     375      N        W      632       43      10   475.0   
2      224   266     263      A        W      880       82      14   480.0   
3      828   838     354      N        E      200       11       3   500.0   
4       48    46      33      N        E      805       40       4    91.5   
..     ...   ...     ...    ...      ...      ...      ...     ...     ...   
317    379   311     138      N        E      325        9       3   700.0   
318    897   451     875      A        E      313      381      20   875.0   
319    217    93     146      A        W       37      113       7   385.0   
320    470   420     332      A        E     1314      131      12   960.0   
321    775   357     249      A        W      408        4       3  1000.0   

    NewLeague  
0           A  
1           N  
2           A  
3           N  
4           N  
..        ...  
317         N  
318         A  
319         A  
320         A  
321         A  

[322 rows x 20 columns]

In [20]:
# dummy data are concatenated
X=pd.concat([X,dummy_data],axis=1)

X






AtBat  Hits  HmRun  Runs  RBI  Walks  Years  CAtBat  CHits  CHmRun  \
0      293    66      1    30   29     14      1     293     66       1   
1      315    81      7    24   38     39     14    3449    835      69   
2      479   130     18    66   72     76      3    1624    457      63   
3      496   141     20    65   78     37     11    5628   1575     225   
4      321    87     10    39   42     30      2     396    101      12   
..     ...   ...    ...   ...  ...    ...    ...     ...    ...     ...   
317    497   127      7    65   48     37      5    2703    806      32   
318    492   136      5    76   50     94     12    5511   1511      39   
319    475   126      3    61   43     52      6    1700    433       7   
320    573   144      9    85   60     78      8    3198    857      97   
321    631   170      9    77   44     31     11    4908   1457      30   

     CRuns  CRBI  CWalks  PutOuts  Assists  Errors  League_N  Division_W  \
0       30    29      14      446       33      20         0           0   
1      321   414     375      632       43      10         1           1   
2      224   266     263      880       82      14         0           1   
3      828   838     354      200       11       3         1           0   
4       48    46      33      805       40       4         1           0   
..     ...   ...     ...      ...      ...     ...       ...         ...   
317    379   311     138      325        9       3         1           0   
318    897   451     875      313      381      20         0           0   
319    217    93     146       37      113       7         0           1   
320    470   420     332     1314      131      12         0           0   
321    775   357     249      408        4       3         0           1   

     NewLeague_N  
0              0  
1              1  
2              0  
3              1  
4              1  
..           ...  
317            1  
318            0  
319            0  
320            0  
321            0  

[322 rows x 19 columns]

In [21]:
# split data
from sklearn.model_selection import train_test_split

X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.25,random_state=42)






In [22]:
print (X_train.shape)
print (X_test.shape)

(241, 19)
(81, 19)


In [23]:
X_train

AtBat  Hits  HmRun  Runs  RBI  Walks  Years  CAtBat  CHits  CHmRun  \
168    315    73      5    23   37     16      4     450    108       6   
148    497   136      7    58   38     26     11    3871   1066      40   
221    576   167      8    89   49     57      4     822    232      19   
227    572   152     18   105   49     65      2     978    249      36   
139    438   103      2    65   32     71      2     440    103       2   
..     ...   ...    ...   ...  ...    ...    ...     ...    ...     ...   
188    279    64      0    31   26     30      1     279     64       0   
71     138    31      8    18   21     38      3     244     53      12   
106    420    95     23    55   58     37      3     646    139      31   
270    303    84      4    35   32     23      2     312     87       4   
102    233    49      2    41   23     18      8    1350    336       7   

     CRuns  CRBI  CWalks  PutOuts  Assists  Errors  League_N  Division_W  \
168     38    46      28      227       15       3         0           1   
148    450   367     241      304      347      10         0           0   
221    132    83      79      325       12       8         1           0   
227    168    91     101      325       13       3         0           1   
139     67    32      71      276        7       9         0           1   
..     ...   ...     ...      ...      ...     ...       ...         ...   
188     31    26      30      107      205      16         1           1   
71      33    32      55      244       21       4         1           0   
106     77    77      61      206       10       7         1           1   
270     39    32      23      179        5       3         1           1   
102    166   122     106      102      132      10         0           0   

     NewLeague_N  
168            0  
148            0  
221            1  
227            0  
139            1  
..           ...  
188            1  
71             1  
106            1  
270            1  
102            0  

[241 rows x 19 columns]

In [24]:
# apply standard scalar
from sklearn.preprocessing import StandardScaler

std=StandardScaler()

X_train=std.fit_transform(X_train)  # z_score=val-mean/std

X_test=std.transform(X_test)  # now just transform






In [25]:
X_train

array([[-0.39752209, -0.56485126, -0.67097354, ..., -0.93178166,
         0.9958592 , -0.93178166],
       [ 0.78043454,  0.79170311, -0.43729634, ..., -0.93178166,
        -1.00415802, -0.93178166],
       [ 1.29174538,  1.45921399, -0.32045774, ...,  1.0732128 ,
        -1.00415802,  1.0732128 ],
       ...,
       [ 0.28206827, -0.09113386,  1.43212127, ...,  1.0732128 ,
         0.9958592 ,  1.0732128 ],
       [-0.47518956, -0.32799256, -0.78781214, ...,  1.0732128 ,
         0.9958592 ,  1.0732128 ],
       [-0.9282498 , -1.08163388, -1.02148934, ..., -0.93178166,
        -1.00415802, -0.93178166]])

In [26]:
# apply linear regression
from sklearn.linear_model import LinearRegression

lin_reg=LinearRegression()

lin_reg.fit(X_train,y_train)

print('Training completed')






Training completed


In [27]:
lin_reg.coef_






array([-3.06135177e+02,  2.58540984e+02,  4.30104638e+01,  2.56873581e-01,
       -2.42897462e+01,  1.03494319e+02, -5.05143118e+01, -4.89744593e+02,
        3.83278121e+02, -3.04654430e+00,  2.92782640e+02,  2.57865392e+02,
       -2.17130332e+02,  5.01331962e+01,  5.35207735e+01, -2.72298561e+01,
        2.53442758e+01, -5.76479218e+01, -1.14922281e+01])

In [28]:
y_pred=lin_reg.predict(X_test)

y_pred

from sklearn.metrics import r2_score
r2_score(y_test,y_pred)

0.36603832997192587

In [29]:
y_pred

array([ 313.45350872,  832.07251524,  314.02121202,  634.77534881,
       1120.01220681,  359.96000328,  457.81774882,  587.77817198,
        149.21527033,  436.63878882,  821.23808042,  346.60347161,
        199.52167756,  916.66904443,  492.25238488,  206.50732829,
        549.37354287,  396.48307356,  571.54376591,  421.92067598,
        386.94917147,  169.95707355,  390.18493905,  910.04111649,
        450.4393045 , 1075.22521517,  441.61964426,  403.09002053,
        901.81420549,  319.70128759,  722.81311549,  453.81275948,
        202.10462793,  502.01740272,  855.49723012,  316.90077465,
        281.26117146,  554.88636547,  818.1775545 ,  322.51297707,
        450.40740552,  397.8934989 ,  433.86981468,  636.94172966,
        330.6838801 , 1271.00752405, 1304.30529402,  352.01461701,
       1076.45861997,  620.31290266,  365.96697238,  135.04589049,
       1417.63033178,  405.63536177,  579.4471586 ,  312.91851041,
        474.8098845 , 1135.23714222,  391.8441266 ,  204.48755

In [30]:
# evaluate model

from sklearn.metrics import mean_squared_error,mean_absolute_error,r2_score

print('mse',mean_squared_error(y_test,y_pred))

print('mae',mean_absolute_error(y_test,y_pred))

print('r2-score test data',r2_score(y_test,y_pred))






mse 119398.48281946743
mae 240.93000429814228
r2-score test data 0.36603832997192587


In [31]:
#calculate r2 score on training data

y_pred_train=lin_reg.predict(X_train)

print('r2-score train data',r2_score(y_train,y_pred_train))

# verry poor data coming so LR is not a good choice
# so non linear pattern is present.






r2-score train data 0.47551324139717344


In [32]:
# to Handle overfitting Lasso and Ridge
from sklearn.linear_model import Lasso

lasso_model=Lasso(max_iter=10000) # iterations are increased otherwise warning was displayed

lasso_model.fit(X_train,y_train)






Lasso(max_iter=10000)

In [33]:
# creating prediction using Lasso
y_pred_lasso=lasso_model.predict(X_test)

y_pred_train_lasso=lasso_model.predict(X_train)





print('r2-score test data lasso',r2_score(y_test,y_pred_lasso))

print('r2-score train data lasso',r2_score(y_train,y_pred_train_lasso))







r2-score test data lasso 0.3835481760030318
r2-score train data lasso 0.4712640255140258


In [34]:
# check lasso model coeff
lasso_model.coef_

array([-310.1213276 ,  264.02397419,   24.11793736,    0.        ,
         -0.        ,   98.00934864,  -71.76154485, -153.66753547,
        113.91763639,    0.        ,  274.79482639,  222.69107914,
       -217.20231461,   50.56545511,   41.72163393,  -23.08815411,
         13.97446905,  -60.24718354,   -1.38805629])

In [35]:
# pass alpha =5  still have overfitting

lasso_model=Lasso(alpha=5,max_iter=10000)

lasso_model.fit(X_train,y_train)






Lasso(alpha=5, max_iter=10000)

In [36]:
y_pred_lasso=lasso_model.predict(X_test)

y_pred_train_lasso=lasso_model.predict(X_train)






In [37]:
print('r2-score test data lasso',r2_score(y_test,y_pred_lasso))

print('r2-score train data lasso',r2_score(y_train,y_pred_train_lasso))






r2-score test data lasso 0.39711842056524327
r2-score train data lasso 0.44004131733187346


In [38]:
lasso_model.coef_ # not important coeff are shrinked to zero


array([-143.03135209,  162.40625728,    4.19254414,    0.        ,
          1.38017514,   59.55433631,  -49.72749203,   -0.        ,
         23.01293789,    0.        ,  122.79269677,  181.3632202 ,
       -104.44072883,   44.51928221,    8.89401036,  -10.23693368,
          8.73394762,  -60.96405042,    0.        ])

In [39]:
# pass alpha =10, increase value of alpha gradually from 1 to 5 to 10 now r2 score good
# when alpha is 0 then its just LR

lasso_model=Lasso(alpha=10,max_iter=10000)

lasso_model.fit(X_train,y_train)

Lasso(alpha=10, max_iter=10000)

In [40]:
# check again both
y_pred_lasso=lasso_model.predict(X_test)

y_pred_train_lasso=lasso_model.predict(X_train)






In [41]:
print('r2-score test data lasso',r2_score(y_test,y_pred_lasso))

print('r2-score train data lasso',r2_score(y_train,y_pred_train_lasso))






r2-score test data lasso 0.3800363892800632
r2-score train data lasso 0.3887552407384661


In [42]:
lasso_model.coef_ # not important coeff are shrinked to zero






array([ -0.        ,  68.13597312,   0.        ,   0.        ,
         0.        ,  25.16897652,  -1.12354281,   0.        ,
        33.27209263,   0.        ,   0.        , 145.23645677,
        -0.        ,  40.14669738,   0.        ,  -5.5420363 ,
         5.20107165, -58.36240638,   0.        ])

In [43]:
X.columns # check in respect to all values

Index(['AtBat', 'Hits', 'HmRun', 'Runs', 'RBI', 'Walks', 'Years', 'CAtBat',
       'CHits', 'CHmRun', 'CRuns', 'CRBI', 'CWalks', 'PutOuts', 'Assists',
       'Errors', 'League_N', 'Division_W', 'NewLeague_N'],
      dtype='object')

In [44]:
# instead of guessing Alpha value, apply LassoCV to find best value of parameter alpha
from sklearn.linear_model import LassoCV

lasso_cv_model=LassoCV(alphas=np.random.randint(1,1000,100),cv=10,max_iter=10000,n_jobs=-1)
# 100 alpha value will be decided among 1000 values randomly
lasso_cv_model.fit(X_train,y_train)

print(lasso_cv_model.alpha_)






7


In [45]:
print (lasso_cv_model.alpha_)
print (lasso_cv_model.coef_)

7
[-52.02406447 101.48956146   0.           0.           0.50937035
  39.29983225 -32.73026073   0.          44.84474852   0.
  43.26104366 162.51442225 -42.13446145  41.70988501   0.
  -7.97212959   7.34997604 -60.64256529   0.        ]


In [46]:
# pass alpha =19
lass_model=Lasso(alpha=19,max_iter=10000)
lasso_model.fit(X_train,y_train)

lass_model.fit(X_train,y_train)

Lasso(alpha=19, max_iter=10000)

In [47]:
y_pred_lasso=lasso_model.predict(X_test)
y_pred_train_lasso=lasso_model.predict(X_train)


In [48]:
print('r2-score test data lasso',r2_score(y_test,y_pred_lasso))
print('r2-score train data lasso',r2_score(y_train,y_pred_train_lasso))
# so 19 is best to keep test n train almost same

r2-score test data lasso 0.3800363892800632
r2-score train data lasso 0.3887552407384661


In [49]:
# apply ridge , lasso n ridge usually not applies on same data, its an expriment here as well
from sklearn.linear_model import Ridge

ridge_model=Ridge(max_iter=10000)

ridge_model.fit(X_train,y_train)






Ridge(max_iter=10000)

In [50]:
y_pred_ridge=ridge_model.predict(X_test)

y_pred_train_ridge=ridge_model.predict(X_train)






In [51]:
print('r2-score test data ridge',r2_score(y_test,y_pred_ridge))

print('r2-score train data ridge',r2_score(y_train,y_pred_train_ridge))






r2-score test data ridge 0.3805191566922891
r2-score train data ridge 0.47223036605110746


In [52]:
# check coeff, all values are present , not shrinked to zero
ridge_model.coef_

array([-298.82701958,  246.8761775 ,   26.21844936,    7.92014656,
         -5.05111208,   98.19364486,  -73.43486658, -207.37526673,
        220.50409988,   24.81948011,  231.685681  ,  188.73867464,
       -212.28493262,   51.26371441,   46.53307108,  -27.79576292,
         22.81415977,  -60.87618628,  -10.24958044])

In [53]:
# recheck to compare lasso coeff with ridge
lasso_model.coef_

array([ -0.        ,  68.13597312,   0.        ,   0.        ,
         0.        ,  25.16897652,  -1.12354281,   0.        ,
        33.27209263,   0.        ,   0.        , 145.23645677,
        -0.        ,  40.14669738,   0.        ,  -5.5420363 ,
         5.20107165, -58.36240638,   0.        ])

In [54]:
# now adding alpha value
from sklearn.linear_model import Ridge

ridge_model=Ridge(alpha=5,max_iter=10000) # applying 254 and running rest all to check

ridge_model.fit(X_train,y_train)






Ridge(alpha=5, max_iter=10000)

In [55]:
y_pred_ridge=ridge_model.predict(X_test)

y_pred_train_ridge=ridge_model.predict(X_train)






In [56]:
print('r2-score test data ridge',r2_score(y_test,y_pred_ridge))

print('r2-score train data ridge',r2_score(y_train,y_pred_train_ridge))






r2-score test data ridge 0.3945820912736484
r2-score train data ridge 0.4579731240377445


In [57]:
ridge_model.coef_






array([-206.83780111,  162.06060783,    3.95529375,   26.07011721,
         18.59319162,   74.07114747,  -76.65344148,  -44.29131193,
        148.60816979,   43.81052382,  133.09011439,  126.50881848,
       -156.58133965,   51.04023725,   36.37737167,  -29.90537726,
         22.92397995,  -62.99329059,  -10.20772412])

In [58]:
lasso_cv_model.coef_

array([-52.02406447, 101.48956146,   0.        ,   0.        ,
         0.50937035,  39.29983225, -32.73026073,   0.        ,
        44.84474852,   0.        ,  43.26104366, 162.51442225,
       -42.13446145,  41.70988501,   0.        ,  -7.97212959,
         7.34997604, -60.64256529,   0.        ])

In [59]:
ridge_model.coef_

array([-206.83780111,  162.06060783,    3.95529375,   26.07011721,
         18.59319162,   74.07114747,  -76.65344148,  -44.29131193,
        148.60816979,   43.81052382,  133.09011439,  126.50881848,
       -156.58133965,   51.04023725,   36.37737167,  -29.90537726,
         22.92397995,  -62.99329059,  -10.20772412])

In [60]:
from sklearn.linear_model import RidgeCV

RidgeCV_model=RidgeCV(alphas=np.random.randint(1,1000,100),cv=10)

RidgeCV_model.fit(X_train,y_train)

print(RidgeCV_model.alpha_)










232


In [61]:
ridge_model.coef_

array([-206.83780111,  162.06060783,    3.95529375,   26.07011721,
         18.59319162,   74.07114747,  -76.65344148,  -44.29131193,
        148.60816979,   43.81052382,  133.09011439,  126.50881848,
       -156.58133965,   51.04023725,   36.37737167,  -29.90537726,
         22.92397995,  -62.99329059,  -10.20772412])

In [62]:
# change alpha to 254
# run all commands

In [66]:
# elastic net
from sklearn.linear_model import ElasticNet
elastic_net=ElasticNet()

elastic_net.fit(X_train,y_train)




y_pred_elastic_net=elastic_net.predict(X_test)

y_pred_train_elastic_net=elastic_net.predict(X_train)












In [67]:
print('r2-score test data elastic_net',r2_score(y_test,y_pred_elastic_net))

print('r2-score train data elastic_net',r2_score(y_train,y_pred_train_elastic_net))






r2-score test data elastic_net 0.3919351475546078
r2-score train data elastic_net 0.37984840332279013
